### PR TITLE
Parallelise our test suite and make errors more readable

### PR DESCRIPTION
**Pull request by alex-zenml**
## Pre-requisites
Please ensure you have done the following:
- [x] I have read the **CONTRIBUTING.md** document.
- [x] If my change requires a change to the documentation, I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Describe changes
- added `pytest-xdist` and `pytest-clarity` dev dependencies (both lightweight)
- add flag into test script to run our tests in parallel over the total # of CPU cores available (for Github Actions, we run on machines that have 2 cores, so we'll get a decent speed bump to our tests)
- add flag into test script for better explainability for error messages (with the `pytest-clarity` plugin package)

### CODE REVIEW

**Comment by htahir1** [comment]
This looks like its giving us a 10-20% boost on testing. Great contribution! Tested locally and it works. Just need to make sure that pre-commit is updated and everybody does `poetry update` or `poetry lock && poetry install` now after merging this. 

Having said that, it does seem that our integration tests use sub-process calls like `pip install` that cause this to fail sporadically. It might make sense to have the parallism only run on the unit tests (i.e. by using poetry run bash scripts/test-coverage-xml.sh unit` and having the script somehow detect that if its only using unit tests then run in parallel mode, and on integration tests run normally?

**Comment by alex-zenml** [comment]
I think that's one solution to have them run only on the unit tests. That said, most of the gains and improvements are going to come from the longer-running integration tests.

I think the way our script is running we can single out the unit tests pretty easily, though. Let me see if I can make it work...

**Comment by alex-zenml** [comment]
We also might want to have a separate `dev-machine` test script which does parallelise everything. I have yet to get any failures running the integration tests in parallel on my local machine.

**Comment by alex-zenml** [comment]
@htahir1 this PR doesn't seem to give us any meaningful speed improvements to our Github Actions times, but at least on a local machine this will give us some benefit.

**Comment by htahir1** [comment]
Is it worth it to introduce dependencies for a local use-case? I think the main benefit would be to get the GitHub Actions working better. Maybe we can investigate a bit how integration parallelism can be achieved. Maybe once we isolate virtualenvs for the tests (ENG-319) this will work as expected? I would keep it open until then

**Comment by alex-zenml** [comment]
@htahir1 I run the integration tests not-infrequently. But then again I can do this all locally without this PR. Maybe others have thoughts. Seems like if we're not gaining anything in the CI, then it means less to introduce this.

**Comment by schustmi** [comment]
I think this will save us valuable time as soon as we start running the integration tests in separate virtual envs as then those `pip install` calls shouldn't interfere with each other. If we don't intend to do that soon maybe we can hold off on this PR until then, but imo we should implement that soon so it wouldn't hurt to merge this PR right away

File: scripts/test-coverage-xml.sh
Comment (schustmi) [inline scripts/test-coverage-xml.sh:17]: Slightest nit: I had to look up both `-n auto` and why we added `-vv` (to enable pytest-clarity), so it might be nice to add a short explanation here. Feel free to ignore if you/others don't think it's necessary

**Review by schustmi — APPROVED** [review]

File: scripts/test-coverage-xml.sh
Comment (alex-zenml) [reply scripts/test-coverage-xml.sh:17]: Do you mean a comment in the script?

**Comment by htahir1** [comment]
Yeah I think its super valuable to have this but again my vote would be to merge it after ENG-319 is done :-)

File: scripts/test-coverage-xml.sh
Comment (schustmi) [reply scripts/test-coverage-xml.sh:17]: Yeah just a one-line comment inside the script that says that it's running on the amount of cpu cores

**Review by AlexejPenner — COMMENTED** [review]
This is cool :) - I like the -vv for readability, will make it easier to understand failures in github actions

**Review by bcdurak — APPROVED** [review]

**Comment by alex-zenml** [comment]
@htahir1 to confirm, this PR shouldn't be merged until whenever we add in the virtualenvs?

**Comment by htahir1** [comment]
Yes @alex-zenml lets wait!

**Comment by strickvl** [comment]
@htahir1 @AlexejPenner assuming these tests all pass, should this now be merged in?

**Comment by strickvl** [comment]
Following discussion with @AlexejPenner I shifted the parallelisation of tests to work only on unit tests (since the virtualenv changes appear not to be compatible with this). Will await feedback on whether to merge this in. @htahir1

**Review by htahir1 — APPROVED** [review]

**Pull request merged**
